src/main/java/de/focusshift/zeiterfassung/security/IndexNameMapSessionRepository.java

```diff
@@ -0,0 +1,74 @@
+package de.focusshift.zeiterfassung.security;
+
+import org.springframework.security.core.Authentication;
+import org.springframework.security.core.context.SecurityContextHolder;
+import org.springframework.security.oauth2.client.authentication.OAuth2AuthenticationToken;
+import org.springframework.security.oauth2.core.oidc.user.OidcUser;
+import org.springframework.security.oauth2.core.user.OAuth2User;
+import org.springframework.session.FindByIndexNameSessionRepository;
+import org.springframework.session.MapSession;
+import org.springframework.session.Session;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+import static java.util.stream.Collectors.toMap;
+
+class IndexNameMapSessionRepository implements FindByIndexNameSessionRepository<Session> {
+
+    private final ConcurrentHashMap<String, Session> sessions = new ConcurrentHashMap<>();
+    // <AttributeName, <AttributeValue, SessionId>>
+    private final ConcurrentHashMap<String, ConcurrentHashMap<String, String>> indexMap = new ConcurrentHashMap<>();
+
+    @Override
+    public Session findById(String id) {
+        return sessions.get(id);
+    }
+
+    @Override
+    public Session createSession() {
+
+        return new MapSession();
+    }
+
+    @Override
+    public void save(Session session) {
+
+        final Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
+        if (authentication instanceof OAuth2AuthenticationToken token) {
+            final OAuth2User oAuth2User = token.getPrincipal();
+            if (oAuth2User instanceof OidcUser oidcUser) {
+                session.setAttribute(PRINCIPAL_NAME_INDEX_NAME, oidcUser.getSubject());
+            }
+        }
+
+        sessions.put(session.getId(), session);
+        for (String attributeName : session.getAttributeNames()) {
+            indexMap.computeIfAbsent(attributeName, k -> new ConcurrentHashMap<>()).put(session.getId(), session.getAttribute(attributeName).toString());
```

Comment:
Do we need to store every attribute, or would we like to narrow down to `PRINCIPAL_NAME_INDEX_NAME`?

Reply:
This implementation will be used by spring session and a session should hold every attribute, otherwise it will not work.